Add Open in Editor support for template paths

federicobond opened the pull request:
#### Description

This adds a setting to configure an Open in editor URL when you click on a template path. Makes it easier to jump directly to the template and edit. Can develop it further and/or add tests if there is interest in the feature.

<img width="569" height="179" alt="Captura de pantalla 2025-09-13 a la(s) 23 04 38" src="https://github.com/user-attachments/assets/b705b1d9-dda5-4519-8ebb-5a8c55737839" />

#### Checklist:

- [x] I have added the relevant tests for this change.
- [x] I have added an item to the Pending section of ``docs/changes.rst``.

federicobond committed:
Add Open in Editor support for template paths

tim-schilling commented on docs/configuration.rst:
```diff
@@ -67,6 +67,14 @@ Toolbar options
+
+  This function is passed a file path and should return a URL to open the file in
+  the editor. For example, to open the file in VS Code, you can use:
+  ``lambda f: f"vscode://file/{f}"``
```
Is this ever not going to be a static prefix with the filename after it? I think if we can remove the lambda, that would be better.

tim-schilling reviewed (commented):
I wonder if it's possible to detect that the project is running in vscode and we could avoid the setting altogether?

tim-schilling commented:
Ah, that's a good reference. I like the idea of having a magic-ish value/constant to specify rather than asking them to write in a lambda for a setting.

federicobond commented:
@tim-schilling updated to use a constant to specify the editor URL format.

Test failures appear unrelated.

federicobond commented on debug_toolbar/utils.py:
```diff
@@ -401,3 +401,28 @@ def is_processable_html_response(response):
+        "idea-remote": "javascript:(()=>{let r=new XMLHttpRequest; r.open('get','http://localhost:63342/api/file/?file={file}&line={line}');r.send();})()",
+        "macvim": "mvim://open/?url=file://{file}&line={line}",
+        "nova": "nova://open?path={file}&line={line}",
+        "pycharm": "pycharm://open?file={file}&line={line}",
```
The line info will be useful for linking traceback lines directly to the correct part of the code.

tim-schilling replied:
Ah, unfortunately pycharm isn't supported from what I can tell.

tim-schilling replied:
This may be a linux only thing. I'm tracking down some help to confirm on windows and mac.

tim-schilling replied:
Yeah, this was a linux only problem. Going to hide these comments.

federicobond committed:
Run editor_url logic when retrieving stats

tim-schilling commented on docs/configuration.rst:
```diff
@@ -67,6 +67,16 @@ Toolbar options
+
+  The editor to use to open file paths from the toolbar.
+
+  Available editors: ``'vscode'``, ``'cursor'``, ``'emacs'``, ``'idea'``,
+  ``'pycharm'``, ``'sublime'``, ``'vscode-insiders'``, ``'vscode-remote'``,
+  ``'vscodium'``, ``'windsurf'``
```
```suggestion
  Available editors:

  * ``'cursor'``
  * ``'emacs'``
  * ``'idea'``
  * ``'pycharm'``
  * ``'sublime'``
  * ``'vscode'``
  * ``'vscode-insiders'``
  * ``'vscode-remote'``
  * ``'vscodium'``
  * ``'windsurf'``
```

Let's make it proper alphabetical order and use some bullets to make it easier to scan for the reader.

tim-schilling reviewed (changes requested):
This is looking good. Marking it as Request Changes to ask for the tests. I think tests that cover `get_editor_url` and the new `get_stats()` of `TemplatesPanel` would be great.

The other stretch thing would be to add is a reference to the new setting in the TemplatesPanel documentation.

federicobond committed:
Add tests for get_editor_url

federicobond committed:
Add tests for open editor url to templates panel

federicobond committed:
Improve EDITOR format in docs

federicobond commented:
I just added the requested tests and updated the docs format.

I would prefer to hold off with documenting this as something that's mostly useful for the requests panel because I believe it's going to be useful in other places too (like stacktraces).

federicobond committed:
Add non-existent editor case to get_editor_url test

federicobond committed:
Split TemplatePanel get_stats test cases

federicobond committed:
Update changelog

tim-schilling reviewed (approved):
I'm good with these changes. I'd like @matthiask to agree (I don't want to merge a new feature unilaterally). He's out on holiday currently, so it may be a few weeks.

federicobond commented:
Rebased against latest `main`.

codingjoe commented on debug_toolbar/utils.py:
```diff
@@ -415,3 +415,28 @@ def get_csp_nonce(request) -> str | None:
+        "windsurf": "windsurf://file/{file}:{line}",
+    }
+    template = formats.get(dt_settings.get_config()["EDITOR"])
+    if template is None:
+        return None
+    return template.format(file=file, line=line)
```
```suggestion
    if template:
        return template.format(file=file, line=line)
```

codingjoe commented on debug_toolbar/utils.py:
```diff
@@ -415,3 +415,28 @@ def get_csp_nonce(request) -> str | None:
+
+
+def get_editor_url(file: str, line: int = 1) -> str | None:
+    formats = {
```
This would be a great application for the all new template strings.

codingjoe commented on debug_toolbar/settings.py:
```diff
@@ -22,6 +22,7 @@ def _is_running_tests():
         "debug_toolbar.panels.profiling.ProfilingPanel",
         "debug_toolbar.panels.redirects.RedirectsPanel",
     },
+    "EDITOR": "vscode",
```
I don't think there should be a default. I'd rather have the feature disabled by default.

tim-schilling replied:
I think defaulting to enabled with vscode is reasonable.

codingjoe replied:
Isn't JetBrains a big sponsor? If we're giving special treatment, let's make it community-friendly.

However, I was going for default OFF. I don't have VSCode installed, so I would get an error.

elineda replied:
I'm for the off by default, no spécial traitment. And if a special traitment there is it should be the most used ide in the django community. Dunno how to find this info.

tim-schilling replied:
@codingjoe as far as I can tell, it doesn't error. It's a no-op in the browser. If we want, we're safe picking _something_ as the default. However, we should consider how to make this setting a bit more obvious. I'll push a commit that adds the note and a link to the setting in the docs.

<img width="1228" height="713" alt="Image" src="https://github.com/user-attachments/assets/d72c27e7-67e9-4c52-9464-10c8ab163e76" />

tim-schilling replied:
Should be resolved in 1d6349f4b6d0381b9ff6c3c5d04ec1b6b0b67bef

codingjoe reviewed (commented):
@federicobond may I be so bold as to propose a slimmer solution?
Instead of hardcoding all the different URIs for all the different tool, why not just give the option to provide a URI as a setting?

If not, this should at least be the fallback. So if the URI is not in the dictionary, whatever string was provided is interpreted as a custom URI template.

tim-schilling committed:
Merge branch 'main' into open-in-editor

tim-schilling committed:
Apply suggestion from @codingjoe

Co-authored-by: Johannes Maron <johannes@maron.family>

tim-schilling committed:
Remove editor default. Add note about configuring editor in panel.
